cedar-policy-validator/src/custom_schema/err.rs

```diff
@@ -0,0 +1,322 @@
+    }
+}
+
+impl<'a> IntoIterator for &'a ParseErrors {
+    type Item = &'a ParseError;
+    type IntoIter = std::slice::Iter<'a, ParseError>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a mut ParseErrors {
+    type Item = &'a mut ParseError;
+    type IntoIter = std::slice::IterMut<'a, ParseError>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter_mut()
+    }
+}
+
+impl Display for ParseErrors {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.first() {
+            Some(first_err) => write!(f, "{first_err}"), // intentionally showing only the first error; see cedar-policy/cedar#326
+            None => write!(f, "{}", Self::DESCRIPTION_IF_EMPTY),
+        }
+    }
+}
+
+impl std::error::Error for ParseErrors {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        self.first().and_then(std::error::Error::source)
+    }
+}
+
+// Except for `.related()`, everything else is forwarded to the first error, if it is present.
+// This ensures that users who only use `Display`, `.code()`, `.labels()` etc, still get rich
+// information for the first error, even if they don't realize there are multiple errors here.
+// See cedar-policy/cedar#326.
+impl Diagnostic for ParseErrors {
+    fn related<'a>(&'a self) -> Option<Box<dyn Iterator<Item = &'a dyn Diagnostic> + 'a>> {
+        // the .related() on the first error, and then the 2nd through Nth errors (but not their own .related())
+        let mut errs = self.iter().map(|err| err as &dyn Diagnostic);
+        errs.next().map(move |first_err| match first_err.related() {
+            Some(first_err_related) => Box::new(first_err_related.chain(errs)),
+            None => Box::new(errs) as Box<dyn Iterator<Item = _>>,
+        })
+    }
+
+    fn code<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
+        self.first().and_then(Diagnostic::code)
+    }
+
+    fn severity(&self) -> Option<miette::Severity> {
+        self.first().and_then(Diagnostic::severity)
+    }
+
+    fn help<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
+        self.first().and_then(Diagnostic::help)
+    }
+
+    fn url<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
+        self.first().and_then(Diagnostic::url)
+    }
+
+    fn source_code(&self) -> Option<&dyn miette::SourceCode> {
+        self.first().and_then(Diagnostic::source_code)
+    }
+
+    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
+        self.first().and_then(Diagnostic::labels)
+    }
+
+    fn diagnostic_source(&self) -> Option<&dyn Diagnostic> {
+        self.first().and_then(Diagnostic::diagnostic_source)
+    }
+}
+
+/// For errors during schema format conversion
+#[derive(Clone, Debug, Error, PartialEq, Eq)]
+pub enum ToJsonSchemaError {
+    /// Error raised when there are duplicate keys
+    #[error("Duplicate keys: `{0}`")]
+    DuplicateKeys(SmolStr, (Loc, Loc)),
```

Comment:
named fields would probably be better than `(SmolStr, (Loc, Loc))` just for clarity.  also in other variants below